### ISSUE TITLE
[RFC] Define services via annotations

### ISSUE THREAD

**Issue by javiereguiluz** [symfony/symfony#21103]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  | 3.3

## Context

One of the goals of Symfony 3.3 is to make the DependencyInjection configuration simpler for newcomers. In Symfony we already have [service autowiring](http://symfony.com/doc/current/components/dependency_injection/autowiring.html) and soon we could also have [getter injection](https://github.com/symfony/symfony/pull/20973). 

## Proposal

However, in my opinion they are missing something. They try to **reduce** the config needed to define services ... but maybe we should instead **remove** the config needed to define services. **We've done this multiple times in Symfony**:

* `@Route` removes the need for `routing.yml | .xml | .php`
* `@Assert` removes the need for `validation.yml | .xml | .php`
* `@ORM` removes the need for `Entity.orm.yml | .xml`

So, why not define a new `@Service` annotation to remove the need for `services.yml | .xml`.

## In practice

It would be very similar to JMSDiExtraBundle. You can configure the services in the PHP file of the related class, without having to define any file. 

#### Alternative 1

```php
use Doctrine\ORM\EntityManager;
use Psr\Log\LoggerInterface;

/** @Service('app.manager.newsletter') */
class NewsletterManager
{
    private $logger;
    private $em;

    /** @InjectParams */
    public function __construct(LoggerInterface $logger, EntityManager $em)
    {
        $this->logger = $logger;
        $this->em = $em;
    }

    // ...
}
```

#### Alternative 2

```php
use Doctrine\ORM\EntityManager;
use Psr\Log\LoggerInterface;

/** @Service('app.manager.newsletter') */
class NewsletterManager
{
    /** Inject('logger') */
    private $logger;
    /** Inject('doctrine.orm.entity_manager') */
    private $em;

    // ...
}
```

#### Alternative 3 (PHP 7+)

```php
use Doctrine\ORM\EntityManager;
use Psr\Log\LoggerInterface;

/** @Service('app.manager.newsletter') */
class NewsletterManager
{
    /** Inject */
    protected function getLogger(): LoggerInterface;

    /** Inject */
    protected function getEm(): EntityManager;

    // ...
}
```

#### Alternative 3 (PHP 5)

```php
use Doctrine\ORM\EntityManager;
use Psr\Log\LoggerInterface;

/** @Service('app.manager.newsletter') */
class NewsletterManager
{
    /** Inject('logger') */
    protected function getLogger();

    /** Inject('doctrine.orm.entity_manager') */
    protected function getEm();

    // ...
}
```

## Comparison

For comparison purposes, this is the traditional Symfony service configuration (which requires 1 PHP file + 1 config file):

```php
use Doctrine\ORM\EntityManager;
use Psr\Log\LoggerInterface;

class NewsletterManager
{
    private $logger;
    private $em;

    public function __construct(LoggerInterface $logger, EntityManager $em)
    {
        $this->logger = $logger;
        $this->em = $em;
    }

    // ...
}
```

```yaml
services:
    app.manager.newsletter:
        class: AppBundle\Manager\NewsletterManager
        arguments: ['@logger', '@doctrine.orm.entity_manager']
        # ...
```

---

And this is the autowiring configuration, which also requires 1 PHP file + 1 config file and the work to do is almost the same as before:

```php
use Doctrine\ORM\EntityManager;
use Psr\Log\LoggerInterface;

class NewsletterManager
{
    private $logger;
    private $em;

    public function __construct(LoggerInterface $logger, EntityManager $em)
    {
        $this->logger = $logger;
        $this->em = $em;
    }

    // ...
}
```

```yaml
services:
    app.manager.newsletter:
        class: AppBundle\Manager\NewsletterManager
        autowire: true
        # ...
```

---

*Mandatory reminder for anything related to autowiring: this feature would be 100% optional and it won't have any impact in your application if you don't want to.*

**Comment by iltar** [comment]
DX wise it might seem nice, but it will also allow people to write a lot of code in a bad way, making it harder for them to decouple it at a later stage. It will surely be nice for new developers as they will have less files to configure, but introduces a path they will go down on which will make them think less in terms of libraries and will start seeing classes as1 services even more.

Services are nothing but a reference to something in a container and people often refer to their class as a service. This feature would make that line even thinner and I'm afraid it would confuse developers even more.

**Comment by GuilhemN** [comment]
At this point, what is the benefit against [DunglasActionBundle](https://github.com/dunglas/DunglasActionBundle) ?
It would add a lot of complexity for a very small gain imo.

**Comment by stof** [comment]
Your second alternative does not work, as the container cannot inject dependencies into private parameters.

Your third and fourth alternatives depend on #20973 (and possibly on #21031), which should be clarified, as it is not yet accepted.

@GuilhemN DunglasActionBundle is about defining controllers differently. This is about services (which might not be controllers themselves)

@javiereguiluz there is one thing to consider there though: defining your services by annotations on the source code directly means that your entire source code now needs to be checked when rebuilding the container. This can have a very negative effect on DX if the container cache is considered as invalidated each time you edit your source code (we used to have this concern with autowired classes at the beginning btw). The implementation must be done in an efficient way (but invalidating the cache when you add the ```@Service`` annotation in a class not using it).
Btw, do we even need ``/** @InjectParams */`` on the constructor ? Can we mark the service as autowired instead (maybe with a property in the ```@Service`` annotation to toggle setter and getter autowiring) ?

**Comment by GuilhemN** [comment]
@stof it is its primary goal but it also works well for other services.
ping @dunglas

**Comment by dknx01** [comment]
Some weeks ago we had such a kind of discussion in our local user group (with the JMSDiExtraBundle).
Most of us don't like the idea very much, because you must look inside (each?) PHP file to find the definition. If you use the services.yml|xml|php file it is easier to find it or at least to know where to look. Also to change the service definition in your local copy or project is easier. 

Beside of this I don't think it is really much easier for newcomers. I think it is easier for them to say "I must have a look into a services.yml|xml|php file. Or search through them."

**Comment by theofidry** [comment]
I'm 👍 for the first alternative, I would also add:

```php
namespace App;

/** @Service */
class NewsletterManager
{
    //...
}
```

which would be equivalent to:

```php
namespace App;

/** @Service('App\NewsletterManager') */
class NewsletterManager
{
    //...
}
```

As it would be more in line with #21133 and auto-wiring.

@javiereguiluz any plan for other services parameters, e.g. auto-wired, private & co.?

**Comment by dunglas** [comment]
I'm not sure that promoting the use of annotations for service definitions is a good idea:

* It encourages to couple the domain (services) with the framework (which provide annotations)
* It makes the usage of internal code (with annotations) different than external random libs (without annotation)
* It breaks the principle of a container dealing with any class from the outside, the idea we try to push with https://github.com/symfony/symfony/projects/2 is that the container becomes (as much as possible) a framework's internal that the user doesn't have to care about, pushing annotation breaks this approach
* What if a lib creator want to support several containers? It will need to add all annotations of the different containers?

IMO, the PHP syntax should be sufficient to specify dependencies of a class (and it's what we do with autowiring, in most cases type hints are enough to inject the good service).
If we really want to be able to use annotations to do that, I think that it should be a standard one like JSR 330 in Java (`javax.inject`), not something specific to Symfony.
Should we propose a PSR or something like that? I'm not sure it's worth it, but it will be better than creating our own stuff.

**Comment by Hanmac** [comment]
@theofidry: i would like it if without name, 
'AppBundle\Service\NewsletterManager' would be turned into "app.newsletter_manager" for example.
it would make it betterr looking compared to the other service names.

**Comment by stof** [comment]
> @stof actually, it is possible to inject dependencies into private properties if the class isn't private, just use an approach similar to what @nicolas-grekas did in #20973

No you cannot. The only way to access private properties is from the class itself (out of our control) or through Reflection (which is bypassing the API of the class, and so a no-go IMO).
#20973 relies on protected visibility, and on methods which allow to hook into their access. A protected property without any setter would require some magic to initialize it (and would mean the class is unusable standalone)

**Comment by stof** [comment]
> What if a lib creator want to support several containers? He'll need to add every annotation provided by the different containers?

@dunglas No, that would be bad as the annotation parser breaks on unknown annotations, which would force to install all containers all the time.
If you want to support multiple containers, the right way is to keep the container configuration in separate files, which is what we already allow in XML and YAML today.
Annotations are not a good fit for packages wanting to have optional deps (this is one of the reasons why FOSUserBundle does not use annotations for the ORM and ODM mapping btw)

> IMO, the PHP syntax should be sufficient to specify dependencies of a class (and it's what we do with autowiring, in most cases type hints are enough to inject the good service).

that's kind of the reason why I asked whether ``@InjectParams`` was really required. IMO, as single ``@Service`` annotation relyinh on autowiring should be enough if we go this way.

**Comment by maryo** [comment]
I agree with @dunglas. Coupling domain objects with ORM is quite OK. You can hardly switch ORM. In case of validation it's mostly OK. The objects are mostly coupled with the validators. In case of routing it's worse but controllers are mostly container aware and depend on Symfony services because it's comfort so it's acceptable tradeoff for me.

But in case of DI... Many if not most classes are reusable and framework independent.

**Comment by Nicofuma** [comment]
Imho, the question of defining services through annotation is not whether or not it is a good idea for libs and bundles to uses them because the answer is no. For me, the libs and bundle question is irrelevant and the only question to ask is whether or not it is useful for the services you write in your application and can make DI usage easier and faster (don't forget: your application is already coupled to the framework, you wo't change it easily). And if I don't really like it for some reasons, I think it is useful and makes a lot of sense with autowiring and the direction DI is taking.

**Comment by hhamon** [comment]
I'm really not a big fan of configuring services with annotations. It seems to be simplier at first sight but it adds a lots of magic from a DX point of view. To me it's even more longer and complicated to write than a simple YAML file.

**Comment by mnapoli** [comment]
Hey @dunglas, there is a start of a discussion of a PSR/standard of `@Inject` (and related annotations) here: https://github.com/container-interop/container-interop/issues/27 (several years old :wink:) Right now the best option would probably be to bring it directly to the FIG though. I still think something like JSR-330 could be useful *for when annotations are used* (which is not always the case).

**Comment by dunglas** [comment]
What would be a nice first step however is to have native annotations in PHP...

**Comment by iltar** [comment]
I don't know why we still don't have those :/

**Comment by Hanmac** [comment]
i don't know if that does target this issue too, but i would like to hear your guys opinions about it.

i often do this in my bundles because i need some object repositories in my other services:

```yml
    app.base_repository:
        abstract:  true
        class: Doctrine\ORM\EntityRepository
        factory: ['@doctrine.orm.entity_manager', getRepository]

    app.xyz_repository:
        parent: app.base_repository
        arguments:
            - AppBundle\Entity\Xyz
```

whould such a ``@Service`` anotation work for this too?
hm or maybe that should be done as param for ``@ORM`` anotations?

or would some kind of dynamic Service Provider that does look for "app.%s_repository" help?

**Comment by stof** [comment]
The annotation would not help here as you need a factory.

Btw, be careful that defining the doctrine repository as a service is incompatible with resetting the entity manager (as the repository would not be replaced by the new repository created by the new entity manager), so I would advice against this.

**Comment by iltar** [comment]
@stof if you have to reset your entity manager, you probably have a bigger issue. I always recommend that if the entity manager is closed, it thus crashed somewhere and it's not safe to continue the normal flow.

**Comment by stof** [comment]
@iltar there is a very valid use case for resetting the manager: long-running process where you want to be able to process the next message in the queue even if the processing of the previous message failed (having to make your consumer process crash all the time is not good).

**Comment by iltar** [comment]
@stof and what about all the (now) un-managed entities floating around everywhere and nowhere? If you try to flush an entity used by another uow, it will most likely cause a bunch of issues and there are probably a dozen more side effects. However, this is slightly going off-topic.

My opinion on service annotations in a class? Bad idea for the aforementioned reasons, I don't think it's worth it. I think autowiring is a better path to go down on and is not too complex to add features to.

**Comment by stof** [comment]
@iltar I'm also clearing the entity manager between processing of different queue messages anyway (otherwise I would be leaking memory). But anyway, this is off-topic for this issue.

**Comment by Guikingone** [comment]
Not a big fan of annotations, if we inject the service using annotations, we link the logic of services to the framework and the class become dependant of the framework logic, i support the vision @dunglas on this part.

Same thinking about getters injection, this way, the logic of the class become dependant of the framework component, what if we want to inject some classes coming from Zend ? Or just moving our service into Zend ? 

I always inject using the constructor, this way, if i need to change the dependencies, i just change the constructor.

**Comment by Nicofuma** [comment]
Lets be honest, you don't change the framework of your project every morning. And when you do change the foundation of your project (yes,, the framework is the foundation of the project) it's a massive task and 2 annotations won't change that much especially if the annotation is only used with autowiring (to not have to write the same 4 lines of yaml for every service)

**Comment by Guikingone** [comment]
For sure, it's just about linking the logic to the framework, not complaining about the changement of framework on a project, i don't change the foundation every morning.

**Comment by javiereguiluz** [comment]
Just for reference, this is how autowiring works in Spring Framework ... and still most developers consider that framework professional, great, etc.

```java
// setter injection
public class Customer
{
    // ...

    @Autowired
    public void setPerson(Person person) {
        this.person = person;
    }
}

// Constructor injection
public class Customer
{
    @Autowired
    public Customer(Person person) {
        this.person = person;
    }

    // ...
}

// Property injection
public class Customer
{
    @Autowired
    private Person person;
    // ...
}
```

The Symfony autowiring feels like "half-done": you want to go fast with autowiring ... but you still need to waste your time with config files. If you want autowiring, you shouldn't have to deal with that.

**Comment by dunglas** [comment]
@javiereguiluz it will not be the case anymore if/when we'll integrate ActionBundle.

**Comment by dunglas** [comment]
About Spring, `@Autowired` is a proprietary and deprecated annotation, for new projects, the standard `@Inject` annotation must be used instead.

**Comment by ianfp** [comment]
I'm in favour of this proposal. I have a project that is very heavy on server-side processing (not a simple CRUD REST backend) with *lots* of services, and I'm finding that maintaining both YAML files and the classes is becoming a hassle. I'm trying out JMSDiExtraBundle and I think it's an improvement, but this proposal seems to have a simpler syntax, which I like.

Specifically, I don't like complex nested annotations because my IDE (PhpStorm) can't syntax-check them:

```
/**
 * @InjectParams({
 *    "preferred" = @Inject("preferred_strategy"),
 *    "fallback" = @Inject("fallback_strategy"),
 * })
 */
public function __construct(Strategy $preferred, Strategy $fallback)
```

In cases where type-hints are not sufficient (eg, when autowiring doesn't work), why not something more like this:

```
/**
 *  @Inject("preferred_strategy", param="preferred")
 *  @Inject("fallback_strategy", param="fallback")
 */
public function __construct(Strategy $preferred, Strategy $fallback)
```

**Comment by weaverryan** [comment]
I've added a possible implementation at #21376

**Comment by javiereguiluz** [comment]
Closing because the related PR (#21376) was closed as "won't fix" and we're exploring other ways to make the DI better.
